### DISCUSSION TITLE
MEAN Stack ainda vale a pena?

### DISCUSSION THREAD

**Discussion by Murillofilho86** [frontendbr/forum#1063] [Perguntas e Respostas]
Há algum tempo atrás falavam que um frontend deveria dominar Mongo, express, angular e node, mas hj, com o crescimento do react e vue... MEAN Stack ainda é válida?

**Comment by helderberto** [comment]
Olá @Murillofilho86, tudo bem?

Sou da opinião que você primeiramente deve dominar Javascript, para assim você ter facilidade ao mudar de framework ou libs. 

No caso de MEAN, caso utilize em seus projetos, a pergunta que fica, está te atendente bem? Caso esteja, você se sente com vontade de aprender novas tecnologias. 
Sugiro testar em projetos pessoais ou contribua em projetos open source, é uma ótima forma de aprender.

Espero ter ajudado, abraços!

**Comment by diessica** [comment]
só tira o A e põe R

[MERN](http://mern.io/)

**Comment by Murillofilho86** [comment]
@helderburato Eu, no momento, estou ficando em HTML, CSS e Javascript. Mas como estou no período final da facul, estou fazendo meu TCC em Node e o que me fez refletir um pouco sobre a área de front. Sou back-end e estou tentando mudar pra front... Tenho pouquíssimo conhecimento em front, só mesmo o que adquiri em uma empresa que estagiei, pegava alguns sites da empresa pra fazer, mas sabe como é, tinha que me virar praticamente sozinho, usava bootstrap e nenhum padrão, acho que a única boa prática na época, era não fazer tudo em tabela... Rsrsrs

**Comment by Murillofilho86** [comment]
@diessica simples e prático, né?! Kkkkk

**Comment by simaodeveloper** [comment]
@diessica ou por V aushaushaus 

[MEVN](https://medium.com/@anaida07/mevn-stack-application-part-1-3a27b61dcae0)

**Comment by kivervinicius** [comment]
Porque focam tanto no Mongo como DB? Ele é tão mil maravilhas assim?

**Comment by houstondapaz** [comment]
@kivervinicius acho legal o foco do MongoDB nessa stack pq ele tb é javascript

**Comment by wilkerfoureaux** [comment]
Mongo é prático, vai dar uma quebra de conceito enorme pra quem tá acostumado ao SELECT * ... tipo "meu primeiro find()" hehe,  mas nada que uma consulta na doc não resolva, tem que estudar, sempre ... 

mas voltando à thread eu aposto muito ainda em stack MEAN
... acredito que o mercado deu uma estabilizada com Angular, pessoal literalmente se adapta bem aos desafios dele
... se o time é pequeno, MEAN ajuda na progressão do projeto e como o @houstondapaz falou e eu pegando o gancho, **é JS**, então tende à *seguir o flow*
... angular 6 saindo do forno aí
... a comunidade é ativa, tem muito conteúdo nas internets pra não te deixar na mão ... MEAN ainda tem muito chão pela frente
... e outro ponto ... posso estar errado, mas eu acho uma stack barata de se startar e manter

**Comment by ninetails** [comment]
~Como futurólogo nos tempos vagos,~ se vc quiser ir pra crista da onda dos hypes, tem o [JAM Stack](https://jamstack.org/), uma stack mais generalizada e sem **único jeito** de se fazer.

A idéia é que seu Markup (HTML) seja estático e cacheável (CDN) e caso precisar de conteúdo _crawleável_, que seja gerado (usando Jekyll, por exemplo); usar JS para o que for dinâmico (bote aqui sua lib/framework); chamando uma API: seu servidor ou pode ser também um serviço, estilo Database as a Service, Firebase, GraphQLs da vida ou mesmo algo serverless estilo AWS Lambda.

**Comment by Murillofilho86** [comment]
@ninetails pesquisei sobre e achei interessante, mas parece um pouco avançado pra quem tá começando, não?

**Comment by ninetails** [comment]
@Murillofilho86 aí voltamos ao que o @helderburato comentou:

> Sou da opinião que você primeiramente deve dominar Javascript, para assim você ter facilidade ao mudar de framework ou libs.

Minha opinião é que a proposta do MEAN stack é tipo ser um bootstrap. Tendo guidelines, vc coda em cima dos frameworks propostos (Angular & Express), agilizando o processo (famoso site pra agência, o pra ontem). Para isso requer que vc saiba apenas Angular e Express (e JS, claro) e como ambas as comunidades são bem representativas, tem muita coisa que vc pode agregar do open source ao seu projeto. Pro TCC, acho que sem problemas.

Porém, dizendo voltado à front, eu diria que se vc sabe bem JS, dá pra encontrar scaffolding/starters (iniciadores de projeto) para praticamente tudo (oficiais: create-react-app, Vue cli). E vc terá a liberdade de usar a ferramenta que mais gostar, ou melhor ainda, usar a ferramenta que mais se adequa ao produto que vc tá fazendo.

Por exemplo: você tem um site institucional de poucas atualizações. Não acharia necessário usar MEAN ou React, famoso canhão pra matar formiga. Mas nem por isso vc precisa também fazer Vanilla. Você pode usar pré/pós processadores pra css e mesmo pro html pra gerar o código estático, deixando mais fácil pra manter, alterar...

Concordo que seria algo avançado para quem tá começando, não deixa de ser verdade. O que eu diria é para não se prender a conhecer somente 2 frameworks (Angular & Express). Não falo pra deixar MEAN de lado, mas sim em não pensar em aprender MEAN como uma finalidade. E ainda diria o mesmo pra Vue e React.

Claro, isso tudo assumindo que vc escolha começar a sua carreira como front. Mas não assuma que precisa saber de tudo isso pra começar, só quis jogar um teaser sobre ser front 😉 Tudo isso se aprende com o tempo.

**Comment by Murillofilho86** [comment]
@ninetails Muito bom... Obrigado pelas dicas, vou correr atras do tempo perdido. rsrs

**Comment by gabrielsimas** [comment]
@ninetails Muito bom... Obrigado pelas dicas, vou correr atras do tempo perdido (2).

**Comment by lguilhermecardoso** [comment]
eu troco o mongo por pg  com ORM =)

**Comment by eliseumds** [comment]
Pela minha experiência, um desenvolvedor frontend deve dominar coisas mais fundamentais e abrangentes que um banco de dados específico como MongoDB ou um UI framework como React.

1. Muitos devs não manjam de **networking** mas é algo que agrega bastante valor: HTTP/S, WS, CORS, DNS, gzip, Cache-Control, ETag e ferramentas para criar e debugar essas coisas
1. **Promises** estão morrendo lentamente em favor de async/await mas ainda há muito software por aí engolindo exceptions pq a galera não entende como funcionam
1. Eu sei que isso é controverso, mas eu cansei de ver bugs gerados pela falta de static typing, então eu recomendo usar **TypeScript**. Fora que a developer experience se torna muito melhor
1. **Webpack** e **Babel** (Babel principalmente para garantir que seu código vai funcionar em todos browsers)
1. Debugging tools (e error-reporting tools como Bugsnag/Sentry/Rollbar, com source maps)
1. **Cloud Services** como AWS, Azure e GCloud. Serviços como autenticação, logging, CDN, API Gateway, notificações (service workers) e file upload, por ex, não são mais exclusivos da área de backend
1. Ferramentas de **continuous integration** como TravisCI/CircleCI/etc. Eventualmente você poderá se tornar responsável pelo time de front-end e terá que configurar linting checks, testes, geração de assets e etc
